clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java

```diff
@@ -247,17 +142,35 @@ public boolean completeExceptionally(Throwable newException) {
      */
     @Override
     public synchronized boolean cancel(boolean mayInterruptIfRunning) {
```

Comment:
Do we still need `synchronized`? It seems to me `CompletableFuture` can deal with thread issue.